### PR TITLE
Peformance improvements in docs/utils.jl

### PR DESCRIPTION
**Pull request by anthonyclays**
Some performance improvements in docutils. I also made a type `ScorePair` to be able to sort 2-tuples - depending on a lexicographic `isless` for tuples feels too pythonic (as there are [multiple ways](https://en.wikipedia.org/wiki/Total_order#Orders_on_the_Cartesian_product_of_totally_ordered_sets) to define an ordening on tuples).

In numbers:

Before:

```
julia> @time Base.Docs.completions(:foo)
  0.036103 seconds (247.24 k allocations: 7.603 MB, 27.98% gc time)
1551-element Array{Any,1}:
 "floor"                 
 "ifloor"                
 "pointer_from_objref"   
...
```

After:

```
julia> @time Base.Docs.completions(:foo)
  0.008606 seconds (36.98 k allocations: 2.219 MB)
1551-element Array{ByteString,1}:
 "floor"                   
 "ifloor"                  
 "pointer_from_objref"     
...
```

### CODE REVIEW

File: base/docs/utils.jl [hunk 1]
```diff
@@ -270,12 +282,17 @@ const builtins = ["abstract", "baremodule", "begin", "bitstype", "break",
 
 moduleusings(mod) = ccall(:jl_module_usings, Any, (Any,), mod)
 
-filtervalid(names) = filter(x->!ismatch(r"#", x), map(string, names))
+isvalid(name::AbstractString) = !ismatch(r"#", name)
```
Comment (tkelman) [inline base/docs/utils.jl]: there's already a function named `isvalid`, probably not a good idea to shadow it here

Comment (anthonyclays) [inline base/docs/utils.jl]: Good catch - I'll rename it to `is_valid_name`.

**Comment by JeffBezanson** [comment]
Cool!

I think it's ok to depend on `isless` being lexicographic order for tuples. `isless` has to be a total order, and there's no other reasonable way to define it for tuples. Hard to imagine us changing it.

**Comment by anthonyclays** [comment]
I agree that having `isless` behave the way it does now is really handy - and there is indeed only one obvious way to define a _total_ order for tuples.

That said, in think the (very few) places that depend on tuple-ordering in Base would be much easier to understand if they didn't depend on that behaviour - for example, the `sanity_check` function in `base/pkg/resolve.jl` sorts a `Vector{Tuple{ByteString,VersionNumber,ByteString}}`. You can't tell right away what sorting a vector of such a tuple actually means.

**Comment by JeffBezanson** [comment]
I don't think it's reasonable to need to define a new type and `isless` method every time you want to sort pairs lexicographically. If you want to be super explicit in a given case, it might be cleaner to pass an `lt` function to `sort`.

We try to avoid lawyerly specifications like "tuples are compared lexicographically, but you're not allowed to depend on that". This isn't POSIX :)

**Comment by anthonyclays** [comment]
Fair enough. I agree that passing a function to `lt` is a very good alternative.

File: base/docs/utils.jl [hunk 2]
```diff
@@ -270,12 +282,17 @@ const builtins = ["abstract", "baremodule", "begin", "bitstype", "break",
...
-accessible(mod::Module) =
-    [names(mod, true, true);
-     map(names, moduleusings(mod))...;
-     builtins] |> unique |> filtervalid
+function accessible(mod::Module)
+    result = ByteString[]
+    append!(result, filter(is_valid_name, map(string, names(mod, true, true))))
+    for inner_mod in moduleusings(mod), sym in names(inner_mod)
+        name = string(sym)
+        if is_valid_name(name); push!(result, name) end
```
Comment (hayd) [inline base/docs/utils.jl]: Small style for similar one-liners is

    is_valid_name(name) && push!(result, name)

But not very important, LGTM.

**Comment by Keno** [comment]
The performance of this is already quite a bit faster on master, so I think the performance analysis here would have to be redone. I'll close this, but if somebody wants to take another look at the performance here, please do.

**Pull request closed without merging**